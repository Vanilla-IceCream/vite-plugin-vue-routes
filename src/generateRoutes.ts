import path from 'path';
import { glob } from 'glob';

import type { PluginOptions } from './types';

export default async function createRoutes(options?: PluginOptions) {
  const dirs = options?.dirs || path.resolve(process.cwd(), 'src/routes');

  const files = await glob(`${dirs}/**/Registry.vue`);

  const paths = [...files]
    .map((file) => {
      const match = file.match(new RegExp(`^${dirs}\\/(.*)\\/Registry\\.vue$`));

      if (match) {
        let path = '/' + match[1];

        // /(group) ->
        path = path.replace(/\/\(.+?\)/g, '');
        if (!path) path += '/';

        // /[...rest] -> /:rest*
        path = path.replace(/\[\.\.\.([^\]]+)\]/g, ':$1*');

        // /[[id]] -> /:id?
        path = path.replace(/\[\[(.+?)\]\]/g, ':$1?');

        // /[id] -> /:id
        path = path.replace(/\[(.+?)\]/g, ':$1');

        return path;
      }

      return null;
    })
    .filter(Boolean);

  const lines: string[] = [];

  files.forEach((item, index) => {
    const mod = `import('${item}')`;
    lines.push(`{ path: '${paths[index]}', component: () => ${mod} },`);
  });

  return `export default () => [${lines.join('')}];`;
}
