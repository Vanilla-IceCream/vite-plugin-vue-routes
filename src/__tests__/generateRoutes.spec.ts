import path from 'path';
import { test, expect } from 'vitest';

import generateRoutes from '../generateRoutes';

test('generateRoutes', async () => {
  const funcStr = await generateRoutes({ dirs: path.resolve(__dirname, '../../examples/src/routes') });

  expect(funcStr).toMatchSnapshot();
});
