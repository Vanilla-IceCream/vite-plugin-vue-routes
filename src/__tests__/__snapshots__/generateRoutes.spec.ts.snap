// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`generateRoutes 1`] = `
"
    export default () => {
      return [{ path: '/products', component: () => import('/Users/shyam_chen/Documents/vite-plugin-vue-routes/examples/src/routes/products/Registry.vue') },{ path: '/hello-world', component: () => import('/Users/shyam_chen/Documents/vite-plugin-vue-routes/examples/src/routes/hello-world/Registry.vue') },{ path: '/', component: () => import('/Users/shyam_chen/Documents/vite-plugin-vue-routes/examples/src/routes/(home)/Registry.vue') },{ path: '/products/:id', component: () => import('/Users/shyam_chen/Documents/vite-plugin-vue-routes/examples/src/routes/products/[id]/Registry.vue') },{ path: '/blog/:info*', component: () => import('/Users/shyam_chen/Documents/vite-plugin-vue-routes/examples/src/routes/blog/[...info]/Registry.vue') },{ path: '/foo', component: () => import('/Users/shyam_chen/Documents/vite-plugin-vue-routes/examples/src/routes/(group)/foo/Registry.vue') },{ path: '/bar', component: () => import('/Users/shyam_chen/Documents/vite-plugin-vue-routes/examples/src/routes/(group)/bar/Registry.vue') },];
    };
  "
`;
